socketshark/receiver.py

```diff
@@ -79,29 +83,35 @@ async def ping_handler(self):
             if wait:
                 wait.cancel()
             if not self._stop:
-                self.shark.log.exception('unhandled exception in ping handler')
+                self.shark.log.debug('redis ping handler cancelled')
         except Exception:
             self.shark.log.exception('unhandled exception in ping handler')
+            self._stop = True
         finally:
-            await self.stop()
+            if self._stop:
+                await self.stop()
```

Comment:
Why do we need these changes?

Reply:
Because of the handler changes mentioned above, the cancel event should not always force the service to shut down. Several tests rely on running `reader(once=True)` multiple times, and we need to distinguish between an expected stop of the ping handler alone, an unexpected error like a missed ping, or a requested shutdown.